src/windows/View.tsx

```diff
@@ -0,0 +1,371 @@
+
+const KEY_CODE_ENTER = 13;
+const KEY_CODE_SPACE = 32;
+
+const DOWN_KEYCODES = [KEY_CODE_SPACE, KEY_CODE_ENTER];
+const UP_KEYCODES = [KEY_CODE_SPACE];
+
+export interface ViewContext {
+    isRxParentAText?: boolean;
+    focusManager?: FocusManager;
+}
+
+// Simple check for the presence of the updated React Native for Windows
+const HasFocusableWindows = (RNW.FocusableWindows !== undefined);
+
+export class View extends ViewCommon implements React.ChildContextProvider<ViewContext>, FocusManagerFocusableComponent {
+    static contextTypes: React.ValidationMap<any> = {
+        isRxParentAText: PropTypes.bool,
+        focusManager: PropTypes.object
+    };
+    context: ViewContext;
+
+    static childContextTypes: React.ValidationMap<any> = {
+        isRxParentAText: PropTypes.bool.isRequired,
+        focusManager: PropTypes.object
+    };
+
+    private _onKeyDown: (e: React.SyntheticEvent<any>) => void;
+    private _onMouseEnter: (e: React.SyntheticEvent<any>) => void;
+    private _onMouseLeave: (e: React.SyntheticEvent<any>) => void;
+    private _onMouseOver: (e: React.SyntheticEvent<any>) => void;
+    private _onMouseMove: (e: React.SyntheticEvent<any>) => void;
+
+    private _focusableElement : RNW.FocusableWindows | null = null;
+
+    private _focusManager: FocusManager;
+    private _isFocusLimited: boolean;
+
+    constructor(props: Types.ViewProps, context: ViewContext) {
+        super(props);
+
+        if (props.restrictFocusWithin || props.limitFocusWithin) {
+            this._focusManager = new FocusManager(context && context.focusManager);
+
+            if (props.limitFocusWithin) {
+                this.setFocusLimited(true);
+            }
+        }
+    }
+
+    componentWillReceiveProps(nextProps: Types.ViewProps) {
+        super.componentWillReceiveProps(nextProps);
+
+        if (!!this.props.restrictFocusWithin !== !!nextProps.restrictFocusWithin) {
+            console.error('View: restrictFocusWithin is readonly and changing it during the component life cycle has no effect');
+        } else if (!!this.props.limitFocusWithin !== !!nextProps.limitFocusWithin) {
+            console.error('View: limitFocusWithin is readonly and changing it during the component life cycle has no effect');
+        }
+    }
+
+    componentDidMount() {
+        super.componentDidMount();
+        if (this._focusManager) {
+            if (this.props.restrictFocusWithin) {
+                this._focusManager.restrictFocusWithin();
+            }
+
+            if (this.props.limitFocusWithin && this._isFocusLimited) {
+                this._focusManager.limitFocusWithin();
+            }
+        }
+    }
+
+    componentWillUnmount() {
+        super.componentWillUnmount();
+        if (this._focusManager) {
+            this._focusManager.release();
+        }
+        // Mixins may have added this
+        if (super.componentWillUnmount !== undefined) {
```

Comment:
>if (super.componentWillUnmount !== undefined) { [](start = 8, length = 47)

Can this be simplified to?:

`if (super.componentWillUnmount) {`

Reply:
Yeah, maybe. I kind of hate the "if (x.method)" syntax, but if that's the customary pattern...

---
In reply to: [159359262](https://github.com/Microsoft/reactxp/pull/404#discussion_r159359262) [](ancestors = 159359262)